idx/metric_definition.go

```diff
@@ -0,0 +1,552 @@
+		return "gauge"
+	case MTypeRate:
+		return "rate"
+	case MTypeCount:
+		return "count"
+	case MTypeCounter:
+		return "counter"
+	case MTypeTimestamp:
+		return "timestamp"
+	default:
+		// case of MTypeUndefined and also default for unknown/corrupted values
+		return ""
+	}
+}
+
+// SetUnit takes a string, interns it in an object store
+// and then uses it to store the unit.
+func (md *MetricDefinition) SetUnit(unit string) {
+	sz, err := IdxIntern.AddOrGetString([]byte(unit), false)
+	if err != nil {
+		log.Errorf("idx: Failed to intern Unit %v. %v", unit, err)
+		internError.Inc()
+		md.Unit = unit
+		return
+	}
+	md.Unit = sz
+}
+
+// SetMetricName interns the MetricName in an
+// object store and stores the addresses of those strings
+// in MetricName.nodes
+func (md *MetricDefinition) SetMetricName(name string) error {
+	nodes := strings.Split(name, ".")
+	md.Name.nodes = make([]uintptr, len(nodes))
+	for i, node := range nodes {
+		// TODO: add error checking? Fail somehow
+		nodePtr, err := IdxIntern.AddOrGet([]byte(node), false)
+		if err != nil {
+			log.Errorf("idx: Failed to intern word in MetricName: %v, %v", node, err)
+			internError.Inc()
+			return fmt.Errorf("idx: Failed to intern word in MetricName: %v, %v", node, err)
+		}
+		md.Name.nodes[i] = nodePtr
+	}
+
+	return nil
+}
+
+// SetTags takes a []string which should contain Key/Value pairs
+// in the form of 'key=value'. It splits up the Key and Value for each
+// item, interns them in the object store, and creates a TagKeyValue
+// for them. It then stores all of these in Tags.
+//
+// The items in the input argument should not contain ';'. Each item
+// is a separate Key/Value pair. Do not combine multiple Key/Value pairs
+// into a single index in the []string.
+func (md *MetricDefinition) SetTags(tags []string) {
+	md.Tags.KeyValues = make([]TagKeyValue, len(tags))
+	sort.Strings(tags)
+	for i, tag := range tags {
+		if tag == "=" || tag == "" {
+			log.Error("idx: SetTags: Empty tag, ignoring: ", tag)
+			invalidTag.Inc()
+			continue
+		}
+		if strings.Contains(tag, ";") {
+			log.Errorf("idx: SetTags: Tag %q has an invalid format, ignoring", tag)
+			invalidTag.Inc()
+			continue
+		}
+		eqPos := strings.Index(tag, "=")
+		if eqPos < 0 {
+			log.Errorf("idx: SetTags: Tag %q has an invalid format, ignoring", tag)
+			invalidTag.Inc()
+			continue
+		}
+		key, err := IdxIntern.AddOrGet([]byte(tag[:eqPos]), true)
+		if err != nil {
+			log.Errorf("idx: SetTags: Failed to intern tag %q, %v", tag, err)
+			internError.Inc()
+			continue
+		} else {
+			md.Tags.KeyValues[i].Key = key
+		}
+
+		value, err := IdxIntern.AddOrGet([]byte(tag[eqPos+1:]), true)
+		if err != nil {
+			log.Errorf("idx: SetTags: Failed to intern tag %q, %v", tag, err)
+			internError.Inc()
+			continue
+		} else {
+			md.Tags.KeyValues[i].Value = value
+		}
+	}
+}
+
+// SetId creates and sets the MKey which identifies a metric
+func (md *MetricDefinition) SetId() {
+	sort.Sort(md.Tags.KeyValues)
```

Comment:
same like above, i think it would be more efficient to sort them as strings if later on they have to be retrieved as strings anyway